testing/xctest.md

```diff
@@ -0,0 +1,37 @@
+
+## Using `final` for `XCTestCase`
+Adding the `final` keyword to your class is an indicator to the reader that this class is not meant to be extended. This makes a lot of sense for `XCTestCase`. There are only a few situations where you would actually want to extend your `XCTestCase` class, but usually if you are doing this then it is probably a code smell that you're doing something wrong.
+
+Using `final` also has [performance implications](https://developer.apple.com/swift/blog/?id=27).
+
+## setUp, tearDown and member variables
+Because of the way XCTestCase manages member variables we have to instantiate and then deallocate them in the `setUp` and `tearDown` fuctions.
+
+For more info on this see [this](https://qualitycoding.org/xctestcase-teardown/)
+
+In light of this, the current best practice is to delcare any members as implicitly unwrapped types, instantiate them in the `setUp` and then set them to nil in the `tearDown`.
+
+
+## Dummys, Fakes, Stubs, Spies and Mocks
+In an attempt to encourage consistency and readability in our tests, we have chosen to follow the vocabulary described in the article [Test Doubles](https://martinfowler.com/bliki/TestDouble.html) when describing the purposes of different objects created for use in tests. A short summary from this article:
+ 
+> - **Dummy** objects are passed around but never actually used. Usually they are just used to fill parameter lists.
+> - **Fake** objects actually have working implementations, but usually take some shortcut which makes them not suitable for production (an [InMemoryTestDatabase](https://martinfowler.com/bliki/InMemoryTestDatabase.html) is a good example).
+> - **Stubs** provide canned answers to calls made during the test, usually not responding at all to anything outside what's programmed in for the test.
+> - **Spies** are stubs that also record some information based on how they were called. One form of this might be an email service that records how many messages it was sent.
+> - **Mocks** are pre-programmed with expectations which form a specification of the calls they are expected to receive. They can throw an exception if they receive a call they don't expect and are checked during verification to ensure they got all the calls they were expecting.
+
+A more thorough discussion of these can be found in [Mocks Aren't Stubs](https://martinfowler.com/articles/mocksArentStubs.html).
+ 
+ There are a number of similar articles online, which we're sure will have various nuances in their explanations and preferences. We valued choosing and documenting one over the specific preferences of any given article.
```

Comment:
here we can maybe have examples of each one, and also the flow diagram that @jorgecmoura suggested